rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,385 @@
+
+## Sample Generated Output
+
+<details>
+<summary><em>Show/Hide Code</em></summary>
+
+```rust
+use crate::OnEvent;
+use crate::render_command::RenderCommand;
+use crate::styles::{Style, StyleProp};
+
+mod my_widget_module {
+    use derivative::Derivative;
+    use crate::{Color, Index, KayakContext, OnEvent, Widget};
+    use crate::styles::Style;
+
+    #[derive(Derivative)]
+    #[derivative(Default, Debug, PartialEq)]
+    pub struct MyWidget {
+        pub id: Index,
+        pub props: MyWidgetProps,
+        pub styles: Option<Style>,
+        #[derivative(Default(value = "None"), Debug = "ignore", PartialEq = "ignore")]
+        pub on_event: Option<crate::OnEvent>,
+        #[derivative(Default(value = "None"), Debug = "ignore", PartialEq = "ignore")]
+        pub children: crate::Children,
+    }
+
+    #[derive(Derivative)]
+    #[derivative(Default, Debug, PartialEq)]
+    pub struct MyWidgetProps {
+        pub name: String,
+        pub fill: Option<Color>,
+    }
+
+    pub struct MyWidgetContext<'a> {
+        context: &'a mut KayakContext,
+        widget: &'a mut MyWidget,
+    }
+
+    impl<'a> MyWidgetContext<'a> {
+        pub fn set_styles(&mut self, styles: Option<Style>) {
+            self.widget.styles = styles;
+        }
+
+        pub fn on_event(&mut self, handler: Option<OnEvent>) {
+            self.widget.on_event = handler;
+        }
+
+        pub fn props(&self) -> &MyWidgetProps {
+            &self.widget.props
+        }
+    }
+
+    impl MyWidget {
+        pub fn generate_context<'a>(&'a mut self, context: &'a mut KayakContext) -> MyWidgetContext<'a> {
+            MyWidgetContext {
+                widget: self,
+                context,
+            }
+        }
+    }
+
+    impl Widget for MyWidget {
+        fn get_id(&self) -> Index {
+            self.id
+        }
+
+        fn focusable(&self) -> bool {
+            true
+        }
+
+        fn set_id(&mut self, id: Index) {
+            self.id = id;
+        }
+
+        fn get_styles(&self) -> Option<Style> {
+            self.styles.clone()
+        }
+
+        fn get_name(&self) -> String {
+            String::from("MyWidget")
+        }
+
+        fn on_event(&mut self, context: &mut KayakContext, event: &mut crate::Event) {
+            let this = self.generate_context(context);
+            if let Some(on_event) = this.widget.on_event.as_mut() {
+                if let Ok(mut on_event) = on_event.0.write() {
+                    on_event(this.context, event);
+                }
+            }
+        }
+
+        fn render(&mut self, context: &mut KayakContext) {
+            Self::render(self.generate_context(context));
```

Comment:
Isn't this a recursive loop?

Reply:
Might be better to say:
```suggestion
            MyWidget::render(self.generate_context(context));
```

Reply:
> Isn't this a recursive loop?

No I meant to call this something else. Maybe `render_internal`?

It's an associated method being called from an instance method, so it shouldn't recurse.

Reply:
Yeah `render_internal` sounds fine.